File: src/main/java/com/viartemev/requestmapper/annotations/spring/extraction/BasePsiAnnotationValueVisitor.java
package com.viartemev.requestmapper.annotations.spring.extraction;

import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiAnnotationMemberValue;
import com.intellij.psi.PsiArrayInitializerMemberValue;
import com.intellij.psi.PsiReferenceExpression;
import org.apache.commons.lang.StringUtils;

import java.util.Collections;
import java.util.List;

public abstract class BasePsiAnnotationValueVisitor implements PsiAnnotationValueVisitor {
    public List<String> visit(PsiAnnotation annotation, String parameter) {
        List<String> result = Collections.emptyList();
        PsiAnnotationMemberValue valueParam = annotation.findAttributeValue(parameter);
        if (valueParam instanceof PsiArrayInitializerMemberValue) {
            result = visitPsiArrayInitializerMemberValue((PsiArrayInitializerMemberValue) valueParam);
        } else if (valueParam instanceof PsiReferenceExpression) {
            result = visitPsiReferenceExpression((PsiReferenceExpression) valueParam);
        } else if (valueParam != null && StringUtils.isNotEmpty(valueParam.getText())) {
            result = visitPsiAnnotationMemberValue(valueParam);
        }
        return result;
    }
}

File: src/main/kotlin/com/viartemev/requestmapper/annotations/spring/extraction/BasePsiAnnotationValueVisitor.kt
package com.viartemev.requestmapper.annotations.spring.extraction

import com.intellij.psi.PsiAnnotation
import com.intellij.psi.PsiArrayInitializerMemberValue
import com.intellij.psi.PsiReferenceExpression
import org.apache.commons.lang.StringUtils

abstract class BasePsiAnnotationValueVisitor : PsiAnnotationValueVisitor {

    fun visit(annotation: PsiAnnotation, parameter: String): List<String> {
        val valueParam = annotation.findAttributeValue(parameter)
        return if (valueParam is PsiArrayInitializerMemberValue) {
            visitPsiArrayInitializerMemberValue(valueParam)
        } else if (valueParam is PsiReferenceExpression) {
            val visitPsiReferenceExpression = visitPsiReferenceExpression((valueParam))
            visitPsiReferenceExpression
        } else if (valueParam != null && valueParam.text.isNotEmpty()) {
            visitPsiAnnotationMemberValue(valueParam)
        } else {
            emptyList()
        }
    }

}
